### PR TITLE
feat: defines internal utils for OAuth client Auth and error handling

### PR DESCRIPTION
**Pull request by bojeil-google**
This defines internal utilities to handle:
- OAuth client authentication
- Standard OAuth response parsing into native Javascript errors.

### CODE REVIEW

File: src/auth/oauth2common.ts [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    bearerToken?: string
+  ) {
+    // Bearer token prioritized higher than basic Auth.
+    if (bearerToken) {
+      opts.headers = opts.headers || {};
+      Object.assign(opts.headers, {
+        Authorization: `Bearer ${bearerToken}}`,
+      });
+    } else if (
+      typeof this.clientAuthentication !== 'undefined' &&
```
[the remark this answers is not in the document]
Comment (bcoe) [reply src/auth/oauth2common.ts:121]: I would just use optional chaining here:

`else if (this?. clientAuthentication?. confidentialClientType === 'basic) {` **wipes hands**.

File: src/auth/oauth2common.ts [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  /**
+   * Applies client authentication on the request's body if request-body
+   * client authentication is selected.
+   *
+   * @param opts The GaxiosOptions whose headers or data are to be modified
+   *   depending on the client authentication mechanism to be used.
+   */
+  private injectAuthenticatedRequestBody(opts: GaxiosOptions) {
+    if (
+      typeof this.clientAuthentication !== 'undefined' &&
```
Comment (bcoe) [inline src/auth/oauth2common.ts:145]: I think we could  use the optional chaining where ever we use this pattern.

File: test/test.oauth2common.ts [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+    };
+    const actualOptions = Object.assign({}, originalOptions);
+    const expectedOptions = Object.assign({}, originalOptions);
+    (expectedOptions.headers as Headers).Authorization = `Bearer ${bearerToken}`;
+
+    handler.testApplyClientAuthenticationOptions(actualOptions, bearerToken);
+    assert.deepStrictEqual(expectedOptions, actualOptions);
+  });
+
+  it('should prioritize bearer token over basic auth', () => {
```
Comment (bcoe) [inline test/test.oauth2common.ts:334]: really happy with the tests that have been added.

File: src/auth/oauth2common.ts [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+import {Crypto, createCrypto} from '../crypto/crypto';
+
+/** List of HTTP methods that accept request bodies. */
+const METHODS_SUPPORTING_REQUEST_BODY = ['PUT', 'POST', 'PATCH'];
+
+/**
+ * OAuth error codes.
+ * https://tools.ietf.org/html/rfc6749#section-5.2
+ */
+export type OAuthErrorCode =
```
[the remark this answers is not in the document]
Comment (bcoe) [reply src/auth/oauth2common.ts:27]: My guess is they're exported for the benefit  of code reuse within the library itself?

**Review by bcoe — COMMENTED** [review]
Left a few initial notes and have some more time slated to dig deeper tomorrow.

In general this structure is looking great.

File: src/auth/oauth2common.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/oauth2common.ts]: In general, the exported types and interfaces here will be re-used within the library but not externally as @bcoe mentioned below.
This type in particular, does not need to be exported. I have removed the export here. Thanks for catching.

File: src/auth/oauth2common.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/oauth2common.ts]: Good point. I ended up using optional chaining as @bcoe suggested. Turned out to be much cleaner :)

File: src/auth/oauth2common.ts (hunk 2 above)
Comment (bojeil-google) [reply src/auth/oauth2common.ts]: Done. I agree this is much cleaner and simpler.

File: test/test.oauth2common.ts (hunk 3 above)
Comment (bojeil-google) [reply test/test.oauth2common.ts:334]: 👍
Code coverage for this file:
```
File                                   | % Stmts | % Branch | % Funcs | % Lines |
  oauth2common.ts                      |     100 |    90.63 |     100 |     100 |
```

File: test/test.oauth2common.ts [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+    it('should inject creds in non-empty x-www-form-urlencoded content', () => {
+      const handler = new TestOAuthClientAuthHandler(reqBodyAuth);
+      const originalOptions: GaxiosOptions = {
+        url: 'https://www.example.com/path/to/api',
+        method: 'POST',
+        headers: {
+          // Handling of headers should be case insensitive.
+          'content-Type': 'application/x-www-form-urlencoded',
+        },
+        data: 'key1=value1&key2=value2',
```
Comment (bcoe) [inline test/test.oauth2common.ts:258]: nit, might consider using:

```
const querystring = require('querystring')
querystring.encode({key1: 'value, k2: 'value2'})
```

Not a requirement.

File: test/test.oauth2common.ts [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+      };
+      const actualOptions = Object.assign({}, originalOptions);
+      const expectedOptions = Object.assign({}, originalOptions);
+      const credentials =
+        `client_id=${encodeURIComponent(reqBodyAuth.clientId)}` +
```
Comment (bcoe) [inline test/test.oauth2common.ts:263]: likewise, might be worth considering using querystring if it makes the code look cleaner, I'm open to template strings if it doesn't seem to look much cleaner.

**Review by bcoe — APPROVED** [review]
This  isa really solid contribution @bojeil-google thank you  👍

File: test/test.oauth2common.ts (hunk 5 above)
Comment (bojeil-google) [reply test/test.oauth2common.ts]: I agree. This is much cleaner.
Done.

File: test/test.oauth2common.ts (hunk 6 above)
Comment (bojeil-google) [reply test/test.oauth2common.ts]: Done here and elsewhere.

**Pull request merged**